packages/flutter_tools/gradle/src/main/groovy/flutter.groovy

```diff
@@ -390,19 +390,63 @@ class FlutterPlugin implements Plugin<Project> {
      * Configures the Flutter plugin dependencies.
      *
      * The plugins are added to pubspec.yaml. Then, upon running `flutter pub get`,
-     * the tool generates a `.flutter-plugins` file, which contains a 1:1 map to each plugin location.
+     * the tool generates a `.flutter-plugins-dependencies` file, which contains a map to each plugin location.
      * Finally, the project's `settings.gradle` loads each plugin's android directory as a subproject.
      */
     private void configurePlugins() {
+        configureLegacyPluginProjects()
         getPluginList().each this.&configurePluginProject
-        getPluginDependencies().each this.&configurePluginDependencies
+        getPluginList().each this.&configurePluginDependencies
+    }
+
+    // TODO(gustl22): Can remove configuring legacy plugins, once #54566 and therefor #48918 is resolved.
+    //  This means all references to `.flutter-plugins` are then removed and apps only depend exclusively on the
+    //  `plugins` property in `.flutter-plugins-dependencies`.
+    //  This cannot be achieved yet as developers may still use an old `settings.gradle`
+    //  which includes all the plugins from the `.flutter-plugins` file, even if not made for Android. 
+    //  The settings.gradle then:
+    //      1) tries to add the android plugin implementation, which does not exist at all, but is also not included
+    //         successfully (which does not throw an error and therefore isn't a problem), or
+    //      2) includes the plugin successfully as a valid android plugin directory exists, even if the surrounding
+    //         flutter package does not support the android platform (see e.g. apple_maps_flutter: 1.0.1). 
+    //         So as it's included successfully it expects to be added as API. This is only possible by taking all 
+    //         plugins into account, which only appear on the `dependencyGraph` and under the `plugins`.
+    //  So in summary the plugins are currently selected from the `dependencyGraph` and filtered then with the
+    //  [doesSupportAndroidPlatform] method instead of just using the `plugins.android` list.
+    private configureLegacyPluginProjects() {
+        File settingsGradle = new File(project.projectDir.parentFile, 'settings.gradle')
+        if(!settingsGradle.exists() || !settingsGradle.text.contains("'.flutter-plugins'")) {
```

Comment:
There is no guarantee that if "settingsGradle.exists()" returns true that the file exists when "settingsGradle.text" is called. A separate process/thread running on the machine can delete it between those two points. This code isn't atomic.

The alternative is to properly handle the exception thrown from "settingsGradle.text"

Reply:
Sorry I have to disagree. I found no place in the flutter repo, where that case is [covered](https://github.com/search?q=repo%3Aflutter%2Fflutter+%22new+File%28%22&type=code). Also it should not throw, if the file does not exist in the start, just return. And in my opinion it is expected to throw, if a file is deleted in exact that moment.

Reply:
I am not sure there can be much disagreement that these actions are not atomic. Reasonable people can disagree about if the effort to detect this case is worth the engineering effort. 

I will leave that decision to @johnmccutchan and if his review will be held until this is fixed.

Reply:
Pointing to existing buggy code doesn't make your argument of "I should add more buggy code" stronger. Here is a wikipedia article explaining the bug in your code in more detail: https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use

Reply:
I'm not arguing that this might not be atomic, but if it's actually necessary.

Regarding the TOCTOU issue, as far as I understood it:
- We don't check any permissions on the file, just if it exists, so it doesn't matter, if the file was changed before or after the `exists` check.
- The followed action is "read only", so it doesn't write to anything critical.

I googled about the issue and I could only found this as reference: https://www.w3schools.com/java/java_files_read.asp

Consequently we should open an issue to refactor the code accordingly, even in the `build.gradle.tmpl`, if this circumstance is considered a flaw.

Addressed in de79906140a225fcfcccb49ca4ba82796c589d76